url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/math/response/Response.java

```diff
@@ -0,0 +1,669 @@
+	 * @param policy the pricing policy
+	 * @param baseline the baseline pricing policy
+	 * @param awareness the person's awareness
+	 * @param sensitivity the person's sensitivity
+	 * @param responseType the response type
+	 * @return the modified Probability Distribution
+	 */
+	public static ProbabilityDistribution respond(ProbabilityDistribution pd,
+			PricingPolicy policy,
+			PricingPolicy baseline,
+			double awareness,
+			double sensitivity,
+			String responseType) {
+		double w = 2;
+		double[] previousHist = pd.getHistogram();
+		double[] newHist = new double[Constants.MIN_IN_DAY];
+		double[] policyArr = policy.getTOUArray();
+		double[] baseArr = baseline.getTOUArray();
+		switch(responseType) {
+		case "None":
+			return pd;
+		case "Optimal":
+			newHist = shiftingOptimal(previousHist, baseArr, policyArr, w * awareness, w * sensitivity);
+			break;
+		case "Normal":
+			newHist = shiftingNormal(previousHist, baseArr, policyArr, w * awareness, w * sensitivity);
+			break;
+		case "Discrete":
+			newHist = shiftingDiscrete(previousHist, baseArr, policyArr, w * awareness, w * sensitivity);
+			break;
+		case "Daily":
+			newHist = shiftingDaily(previousHist, baseArr, policyArr, awareness, sensitivity);
+			// Clean up NaNs
+			for(int i = 0; i < newHist.length; i++) {
+				if(Double.isNaN(newHist[i])) {
+					newHist[i] = 0.0;
+				}
+			}
+			break;
+		default:
+			return pd;
+		}
+		ProbabilityDistribution retPd = new Histogram(newHist);
+		return retPd;
+	}
+
+	private static double[] shiftingNormal (double[] values,
+			double[] basicScheme, double[] newScheme,
+			double awareness, double sensitivity) {
+		double[] result = Arrays.copyOf(values, values.length);
+		PricingVector pricingVector = new PricingVector(basicScheme, newScheme);
+		IncentiveVector inc = new IncentiveVector(basicScheme, newScheme);
+		if (pricingVector.getPricings().size() > 1)
+			for (Incentive incentive: inc.getIncentives())
+				result = movingAverage(result, incentive, awareness, sensitivity);
+		return result;
+	}
+
+	private static double[] shiftingDiscrete (double[] values,
+			double[] basicScheme, double[] newScheme,
+			double awareness, double sensitivity) {
+		double[] result = Arrays.copyOf(values, values.length);
+		PricingVector pricingVector = new PricingVector(basicScheme, newScheme);
+		if (pricingVector.getPricings().size() > 1)
+			result = discreteAverage(result, pricingVector, awareness, sensitivity);
+		return result;
+	}
+
+	private static double[] shiftingOptimal (double[] values,
+			double[] basicScheme, double[] newScheme,
+			double awareness, double sensitivity) {
+		double[] result = Arrays.copyOf(values, values.length);
+		PricingVector pricingVector = new PricingVector(basicScheme, newScheme);
+		if (pricingVector.getPricings().size() > 1)
+			result = discreteOptimal(result, pricingVector, awareness, sensitivity);
+		return result;
+	}
+
+	private static double[] movingAverage (double[] values, Incentive incentive,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 39. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.